### PR TITLE
Run Datacomp at scale

### PR DESCRIPTION
**Pull request by NielsRogge**
This PR includes updates made to the Fondant code base to make it run at scale for Datacomp

### CODE REVIEW

File: components/load_from_hf_hub/src/main.py [hunk 1]
```diff
@@ -44,12 +48,27 @@ def load(self) -> dd.DataFrame:
...
+            npartitions = self.n_rows_to_load // partition_length
+            dask_df = dask_df.head(self.n_rows_to_load, npartitions=npartitions)
+            dask_df = dd.from_pandas(dask_df, npartitions=npartitions)
+            # .reset_index(drop=True) # will reset it from 0 for every partition
+
+        # Set monotonically increasing index
+        logger.info("Setting the index...")
+        dask_df["id"] = 1
+        dask_df["id"] = dask_df.id.cumsum()
+        dask_df = dask_df.set_index("id", sort=True)
```
Comment (NielsRogge) [inline components/load_from_hf_hub/src/main.py:71]: This is based on https://stackoverflow.com/questions/47571715/dask-create-strictly-increasing-index

File: components/filter_image_resolution/fondant_component.yaml [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
 description: Component that filters images based on minimum size and max aspect ratio
-image: ghcr.io/ml6team/filter_image_resolution:latest
+image: ghcr.io/ml6team/filter_image_resolution:f3f3925b8e8f634e2978e5c7fcefa72c53baba7c
 
 consumes:
   image:
     fields:
       width:
-        type: int16
+        type: int64
```
Comment (PhilippeMoussalli) [inline components/filter_image_resolution/fondant_component.yaml:9]: int16 maxes out at 32.767 which is why you may be having issues if your image's width was 36.000 
However, I think this should be resolved if you move to `int32`

File: components/filter_image_resolution/Dockerfile [hunk 3]
```diff
@@ -3,15 +3,16 @@ FROM --platform=linux/amd64 python:3.8-slim
...
+    apt-get install graphviz -y
 
 # Install requirements
 COPY requirements.txt /
 RUN pip3 install --no-cache-dir -r requirements.txt
 
 # Install Fondant
 # This is split from other requirements to leverage caching
-ARG FONDANT_VERSION=main
+ARG FONDANT_VERSION=f3f3925b8e8f634e2978e5c7fcefa72c53baba7c
```
Comment (PhilippeMoussalli) [inline components/filter_image_resolution/Dockerfile:15]: @GeorgesLorre do we still need to revert the docker image to `main` or do they get automatically tagged as `dev` during merge?

File: components/load_from_hf_hub/fondant_component.yaml [hunk 4]
```diff
@@ -23,4 +23,9 @@ args:
   n_rows_to_load:
     description: Optional argument that defines the number of rows to load. Useful for testing pipeline runs on a small scale
     type: int
-    default: None
\ No newline at end of file
+    default: None
+  dataset_length:
```
Comment (PhilippeMoussalli) [inline components/load_from_hf_hub/fondant_component.yaml:27]: I would avoid having the user specify the length of the dataset and read it from the hf metadata directly

File: examples/pipelines/datacomp/components/download_images/src/main.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
+"""
```
Comment (PhilippeMoussalli) [inline examples/pipelines/datacomp/components/download_images/src/main.py:1]: What are the reasons for duplicating this component and not making the changes directly to the one in the registry?

File: components/filter_image_resolution/Dockerfile [hunk 6]
```diff
@@ -3,15 +3,16 @@ FROM --platform=linux/amd64 python:3.8-slim
 # System dependencies
 RUN apt-get update && \
     apt-get upgrade -y && \
-    apt-get install git -y
+    apt-get install git -y && \
+    apt-get install graphviz -y
```
Comment (PhilippeMoussalli) [inline components/filter_image_resolution/Dockerfile:7]: You can omit `graphviz` for now, we'll tackle this in #331

File: pyproject.toml [hunk 7]
```diff
@@ -46,6 +46,7 @@ dask = {extras = ["dataframe"], version = ">= 2023.4.1"}
 importlib-resources = { version = ">= 1.3", python = "<3.9" }
 jsonschema = ">= 4.18"
 pyarrow = ">= 11.0.0"
+graphviz = ">= 0.20.1"
```
Comment (PhilippeMoussalli) [inline pyproject.toml:49]: same comment here as above, maybe for testing you can just have it in the `requirements.txt` for now

File: scripts/build_components.sh [hunk 8]
```diff
@@ -75,7 +75,7 @@ for dir in "${components_to_build[@]}"; do
 
   echo "Updating the image version in the fondant_component.yaml with:"
   echo "${full_image_names[0]}"
-  sed -i "s|^image: .*|image: ${full_image_names[0]}|" fondant_component.yaml
+  sed -i '' "s|^image: .*|image: ${full_image_names[0]}|" fondant_component.yaml
```
Comment (PhilippeMoussalli) [inline scripts/build_components.sh:78]: what is the extra `''` for?

File: src/fondant/data_io.py [hunk 9]
```diff
@@ -92,6 +93,9 @@ def _load_subset(self, subset_name: str, fields: t.List[str]) -> dd.DataFrame:
 
         subset_df = dd.read_parquet(remote_path, columns=fields)
 
+        logger.info(f"First few rows of subset {subset_name}:")
```
Comment (PhilippeMoussalli) [inline src/fondant/data_io.py:96]: I guess this is only for debugging and can be omitted

File: src/fondant/data_io.py [hunk 10]
```diff
@@ -124,11 +128,16 @@ def load_dataframe(self) -> dd.DataFrame:
...
+        logging.info(f"Loading index...")
         dataframe = self._load_index()
+
         for name, subset in self.component_spec.consumes.items():
             fields = list(subset.fields.keys())
             subset_df = self._load_subset(name, fields)
             # left joins -> filter on index
+            # make sure that dataframe has same number of partitions
+            # as subset
+            dataframe = dataframe.repartition(npartitions=subset_df.npartitions)
```
Comment (PhilippeMoussalli) [inline src/fondant/data_io.py:140]: I'll have a look at this, eventually we can maybe even flag it as an issue to the `Dask` repo

File: src/fondant/data_io.py [hunk 11]
```diff
@@ -196,6 +206,12 @@ def write_dataframe(self, dataframe: dd.DataFrame) -> None:
 
         dataframe.index = dataframe.index.rename("id").astype("string")
 
+        # logging.info("Visualizing task graph...")
```
Comment (PhilippeMoussalli) [inline src/fondant/data_io.py:209]: If I have custom code that I use for testing but don't want to commit I usually use 
```
git add -p <filename>
```
which allows you to only commit chunks of the file, can be also done in the IDE. 

Then if I want to switch to another branch and not lose the testing code, I stash the changes under a descriptive name

```
git stash save datacomp-scale-test
```

You can then fetch it back using 
```
git stash apply <stash_id>
```

File: src/fondant/schemas/common.json [hunk 12]
```diff
@@ -7,6 +7,7 @@
         "int8",
         "int16",
         "int32",
+        "int64",
```
Comment (PhilippeMoussalli) [inline src/fondant/schemas/common.json:10]: This can be kept, I would just check that previous component can run at `int32` to reduce memory footprint

**Review by PhilippeMoussalli — COMMENTED** [review]
Thanks Niels! 
I'll try to reproduce the steps and investigate the merging issue

File: scripts/build_components.sh (hunk 8 above)
Comment (NielsRogge) [reply scripts/build_components.sh:78]: This was required to run the script on MacOS: https://stackoverflow.com/questions/29081799/sed-1-invalid-command-code-f

**Comment by NielsRogge** [comment]
Closing this PR in favor of a more up-to-date branch.

**Pull request closed without merging**